github/ratelimit.go

```diff
@@ -0,0 +1,219 @@
+func NewRateLimitTransport(rt http.RoundTripper, logger log.Logger) *RateLimitTransport {
+	return &RateLimitTransport{
+		transport: rt,
+		logger:    logger,
+	}
+}
+
+// RoundTrip executes a single HTTP transaction, returning a Response for the provided Request.
+// If the request hitted an API RateLimit or Abuse, it will return an ErrorRateLimit
+// and it no longer process any further Requests until the Limit has been expired.
+func (rt *RateLimitTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	// Make Requests serially
+	rt.lock()
+	defer rt.unlock()
+
+	if time.Now().Before(rt.lockedUntil) {
+		rt.logger.Infof("rate limit reached, sleeping until %s", rt.lockedUntil)
+		time.Sleep(rt.lockedUntil.Sub(time.Now()))
+	}
+
+	resp, err := rt.transport.RoundTrip(req)
+	if err != nil {
+		return resp, err
+	}
+
+	if errRateLimit := checkResponseRateLimit(resp, rt.logger); errRateLimit != nil {
+		rt.lockedUntil = errRateLimit.when()
+		rt.logger.Errorf(errRateLimit, "locking transport for %s", errRateLimit.when().Sub(time.Now()))
+		return nil, errRateLimit
+	}
+
+	return resp, nil
+}
+
+func (rt *RateLimitTransport) lock() {
+	rt.mutex.Lock()
+}
+
+func (rt *RateLimitTransport) unlock() {
+	rt.mutex.Unlock()
+}
+
+// checkRateLimit checks the API response and returns a whener error if a rate limit was found:
+// - *ErrRateLimit is returned when the request failed because of a RateLimit
+//    https://developer.github.com/v4/guides/resource-limitations/#rate-limit
+// - *ErrAbuseRateLimit is returned when the request triggered a GitHub abuse detection mechanism
+//    https://developer.github.com/v3/#abuse-rate-limits
+func checkResponseRateLimit(resp *http.Response, logger log.Logger) whener {
+	if err := asErrRateLimit(resp); err != nil {
+		return err
+	}
+
+	if err := asErrAbuseRateLimit(resp, logger); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// ErrRateLimit is returned when a request failed because of a RateLimit
+// https://developer.github.com/v4/guides/resource-limitations/#rate-limit
+type ErrRateLimit struct {
+	errRetryLater
+}
+
+func (e *ErrRateLimit) Error() string {
+	return fmt.Sprintf("API rate limit exceeded; %s", e.errRetryLater.Error())
+}
+
+// ErrAbuseRateLimit is returned when a request triggers any GitHub abuse detection mechanism
+// https://developer.github.com/v3/#abuse-rate-limits
+type ErrAbuseRateLimit struct {
+	errRetryLater
+}
+
+func (e *ErrAbuseRateLimit) Error() string {
+	return fmt.Sprintf("abuse detection mechanism triggered; %s", e.errRetryLater.Error())
+}
+
+type errRetryLater struct {
+	retryAfter time.Time
+}
+
+func (e *errRetryLater) Error() string {
+	return fmt.Sprintf("retry after %s", e.retryAfter)
+}
+
+func (e *errRetryLater) when() time.Time {
+	return e.retryAfter
+}
+
+type whener interface {
+	error
+	when() time.Time
+}
+
+/*
+An apiErrorResponse reports one or more errors caused by an API request.
+GitHub API docs: https://developer.github.com/v3/#client-errors
```

Comment:
In the GitHub v4 API I have observed that if the JSON payload is OK, but the query graphql string is not, it returns a `200 OK` status with a body `{"errors":[{"message":"Unexpected end of document"}]}`.

Reply:
mmm... not sure about how that relates to this PR.
In that case, the response you describe won't be considered as an `Abuse`, [because it won't be `http.StatusForbidden`](https://github.com/src-d/metadata-retrieval/pull/57/files#diff-68a4c629ebdc3254db9da66296906b7aR164), nor a `RateLimit` because even being `http,StatusOK`, [it won't have the `X-RateLimit-Remaining` header set to `0`](https://github.com/src-d/metadata-retrieval/pull/57/files#diff-68a4c629ebdc3254db9da66296906b7aR202).

Reply:
I just wrote it wrt to the comment...that v4 client errors at least at the moment can be different from v3.

Reply:
I'd say that returning errors as `200` is part of the consecuences of the [graphql specs](https://graphql.github.io/graphql-spec/June2018/#sec-Errors)
I didn't read it in depth, but I'd interpret it in that way.

Anyhow, per @kyrcha suggestion via slack, I added this: 370406b